### PR TITLE
Barbarian Potions + Roe/Caviar 

### PR DESCRIPTION
**Pull request by TastyPumPum**
Adds the cutting of leaping fish to the cook command. 
Moved it into its own command and activity to keep it tidy.
Used the same rates as crushing as its a similar method.

Additionally added barbarian mixes to the /mix and /decant commands
Also tidied up the mixables using item: getOSItem()

-   [x] I have tested all my changes thoroughly.

### CODE REVIEW

File: src/tasks/minions/cutLeapingFishActivity.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		let caviarCreated = 0;
+		let roeChance = 0;
+		let roeCreated = 0;
+		let fishOffcutsChance = 0;
+		let fishOffcutsCreated = 0;
+
+		if (BarbarianFish.name === 'Cut leaping sturgeon') {
+			caviarChance = 1.25 * currentLevel;
+			fishOffcutsChance = (5 / 6) * 100;
+			if (caviarChance > 100) caviarChance = 100;
```
Comment (Superwhammybar) [inline src/tasks/minions/cutLeapingFishActivity.ts:27]: Line 25 and 27 can be condensed to `caviarChance = clamp(1.25 * currentLevel, 0, 100)`

File: src/tasks/minions/cutLeapingFishActivity.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+				if (percentChance(fishOffcutsChance)) {
+					fishOffcutsCreated += 1;
+				}
+			}
+		}
+
+		if (BarbarianFish.name === 'Cut leaping salmon') {
+			roeChance = 1.25 * currentLevel;
+			fishOffcutsChance = (3 / 4) * 100;
+			if (roeChance > 100) roeChance = 100;
```
Comment (Superwhammybar) [inline src/tasks/minions/cutLeapingFishActivity.ts:40]: Line 41 and 43 can be condensed to `roeChance = clamp(1.25 * currentLevel, 0, 100)`

File: src/tasks/minions/cutLeapingFishActivity.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+				if (percentChance(fishOffcutsChance)) {
+					fishOffcutsCreated += 1;
+				}
+			}
+		}
+
+		if (BarbarianFish.name === 'Cut leaping trout') {
+			roeChance = 0.67 * currentLevel;
+			fishOffcutsChance = (1 / 2) * 100;
+			if (roeChance > 100) roeChance = 100;
```
Comment (Superwhammybar) [inline src/tasks/minions/cutLeapingFishActivity.ts:54]: Line 57 and 59 can be condensed to `roeChance = clamp(0.67 * currentLevel, 0, 100)`

**Review by Superwhammybar — CHANGES_REQUESTED** [review]

File: src/tasks/minions/cutLeapingFishActivity.ts (hunk 3 above)
Comment (TastyPumPum) [reply src/tasks/minions/cutLeapingFishActivity.ts]: Updated with your changes, thanks :)

File: src/lib/types/minions.ts [hunk 4]
```diff
@@ -175,6 +175,11 @@ export interface HerbloreActivityTaskOptions extends ActivityTaskOptions {
 	zahur: boolean;
 }
 
+export interface CutLeapingFishActivityTaskOptions extends ActivityTaskOptions {
+	fishName: string;
```
Comment (gc) [inline src/lib/types/minions.ts:179]: never put names in task options. use IDs only.

File: src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		)}.`;
+	}
+
+	const finalCost = requiredItems.multiply(quantity);
+	if (!user.owns(finalCost)) {
+		return `You don't own: ${finalCost}.`;
+	}
+	await user.removeItemsFromBank(finalCost);
+
+	updateBankSetting('herblore_cost_bank', finalCost);
```
Comment (gc) [inline src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts:65]: is this meant to be herblore?

File: src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+	const maxTripLength = calcMaxTripLength(user, 'Herblore');
+
+	if (!quantity) quantity = Math.floor(maxTripLength / timeToCutSingleItem);
+
+	const baseCost = new Bank(barbarianFish.inputItems);
+
+	const maxCanDo = user.bankWithGP.fits(baseCost);
+	if (maxCanDo === 0) {
+		return `You don't have enough supplies to cut even one of this item!\nTo ${barbarianFish.name}, you need to have ${baseCost}.`;
```
Comment (gc) [inline src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts:43]: Typo

File: src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+			_leapingFish.aliases.some(alias => stringMatches(alias, name))
+	);
+
+	if (!barbarianFish) return 'That is not a valid fish to cut.';
+
+	let requiredItems = new Bank(barbarianFish.inputItems);
+
+	let timeToCutSingleItem = barbarianFish.tickRate * Time.Second * 0.6;
+
+	const maxTripLength = calcMaxTripLength(user, 'Herblore');
```
Comment (gc) [inline src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts:35]: herb?

File: src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	await addSubTaskToActivityTask<CutLeapingFishActivityTaskOptions>({
+		fishName: barbarianFish.name,
+		userID: user.id,
+		channelID: channelID.toString(),
+		quantity,
+		duration,
+		type: 'CutLeapingFish'
+	});
+
+	return `${user.minionName} is now cutting ${quantity} ${barbarianFish.name}, it'll take around ${formatDuration(
```
Comment (gc) [inline src/mahoji/lib/abstracted_commands/cutLeapingFishCommand.ts:76]: put `x` after the quantity

File: src/tasks/minions/cutLeapingFishActivity.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		xpReceived += 10 * roeCreated;
+		xpReceived += 15 * caviarCreated;
+
+		const xpRes = await user.addXP({
+			skillName: SkillsEnum.Cooking,
+			amount: xpReceived,
+			duration
+		});
+
+		let str = `${user}, ${user.minionName} finished ${barbarianFish.name}. ${xpRes}\n\n You received: ${loot}.`;
```
Comment (gc) [inline src/tasks/minions/cutLeapingFishActivity.ts:78]: typo

**Review by gc — CHANGES_REQUESTED** [review]
The ones listed as typo are because you put the verb in the name of the fish and so it looks weird in the code - which doesnt really make sense. id suggest you can just remove the name and inputItems, and instead put  `item: getOSItem("name")`, that will give you an id and name. and then you write `finished cutting ${fish.name}`

**Pull request merged**